File: 42. Trapping Rain Water/42. Trapping Rain Water.go
func trap(height []int) int {
    if len(height) <= 2 {
        return 0
    }
    
    left, right := 0, len(height) - 1
    leftMax, rightMax := 0, 0
    water := 0
    
    for left < right {
        if height[left] < height[right] {
            if height[left] >= leftMax {
                leftMax = height[left]
            } else {
                water += leftMax - height[left]
            }
            left++
        } else {
            if height[right] >= rightMax {
                rightMax = height[right]
            } else {
                water += rightMax - height[right]
            }
            right--
        }
    }
    
    return water
}

File: 42. Trapping Rain Water/42. Trapping Rain Water.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let n = height.len();
        if n < 3 {
            return 0;
        }
        
        let mut left = 0;
        let mut right = n - 1;
        let mut left_max = 0;
        let mut right_max = 0;
        let mut water = 0;
        
        while left < right {
            if height[left] < height[right] {
                if height[left] >= left_max {
                    left_max = height[left];
                } else {
                    water += left_max - height[left];
                }
                left += 1;
            } else {
                if height[right] >= right_max {
                    right_max = height[right];
                } else {
                    water += right_max - height[right];
                }
                right -= 1;
            }
        }
        
        water
    }
}